acceptance/tests/connect/local_rate_limit_test.go

```diff
@@ -0,0 +1,126 @@
+	rateLimitMap := map[string]int{
+		"http://localhost:1234":             2,
+		"http://localhost:1234/exact":       3,
+		"http://localhost:1234/prefix-test": 4,
+		"http://localhost:1234/regex":       5,
+	}
+
+	opts := newRateLimitOptions(t, ctx)
+
+	// Ensure that all requests from static-client to static-server succeed.
+	for addr, rps := range rateLimitMap {
+		opts.rps = rps
+		assertRateLimits(t, opts, addr)
+	}
+
+	// Apply local rate limiting to the static-server
+	writeCrd(t, connHelper, "../fixtures/cases/local-rate-limiting")
+
+	// Ensure that going over the limit causes the static-server to apply rate limiting and
+	// reply with 429 Too Many Requests
+	opts.enforced = true
+	for addr, reqPerSec := range rateLimitMap {
+		opts.rps = reqPerSec
+		assertRateLimits(t, opts, addr)
+	}
+}
+
+func assertRateLimits(t *testing.T, opts *assertRateLimitOptions, curlArgs ...string) {
+	t.Helper()
+
+	args := []string{"exec", opts.resourceType + opts.sourceApp, "-c", opts.sourceApp, "--", "curl", opts.curlOpts}
+	args = append(args, curlArgs...)
+
+	// This check is time sensitive due to the nature of rate limiting.
+	// Run the entire assertion in a retry block and on each pass:
+	// 1. Send the exact number of requests that are allowed per the rate limiting configuration
+	//    and check that all the requests succeed.
+	// 2. Send an extra request that should exceed the configured rate limit and check that this request fails.
+	// 3. Make sure that all requests happened within the rate limit enforcement window of one second.
+	retry.RunWith(opts.retryTimer, t, func(r *retry.R) {
+		// Make up to the allowed numbers of calls in a second
+		t0 := time.Now()
+		for i := 0; i < opts.rps; i++ {
+			output, err := k8s.RunKubectlAndGetOutputE(t, opts.k8sOpts, args...)
+			require.NoError(r, err)
+			require.Contains(r, output, opts.successOutput)
+		}
+
+		// Exceed the configured rate limit.
+		output, err := k8s.RunKubectlAndGetOutputE(t, opts.k8sOpts, args...)
+		if opts.enforced {
+			require.Error(r, err)
+			require.Contains(r, output, opts.rateLimitOutput, "request was not rate limited")
+		} else {
+			require.NoError(r, err)
+			require.Contains(r, output, opts.successOutput, "request was not successful")
+		}
+		require.True(r, time.Since(t0) < time.Second, "failed to make all requests within one second window")
+	})
+}
+
+type assertRateLimitOptions struct {
+	resourceType    string
+	successOutput   string
+	rateLimitOutput string
+	k8sOpts         *terratestK8s.KubectlOptions
+	sourceApp       string
+	rps             int
+	enforced        bool
+	retryTimer      *retry.Timer
+	curlOpts        string
+}
+
+func newRateLimitOptions(t *testing.T, ctx environment.TestContext) *assertRateLimitOptions {
+	return &assertRateLimitOptions{
+		resourceType:    "deploy/",
+		successOutput:   "hello world",
+		rateLimitOutput: "curl: (22) The requested URL returned error: 429",
```

Comment:
I might even consider changing this to just contain `429` in case any part of this string were to change in the future, 429 would be the most likely part to stay consistent

Reply:
It's possible that this could change but I think it's unlikely :shrug:. If curl changes its output format/messages we'll likely have to change things elsewhere. I followed what was used in the `connhelper` pkg for testing failed connections: https://github.com/hashicorp/consul-k8s/blob/3d3bd70b7168c346e55be0dbfdae7175111ad25a/acceptance/framework/connhelper/connect_helper.go#L417-L419